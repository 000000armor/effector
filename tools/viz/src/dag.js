//@flow

//$todo
import {line as lined3, curveCatmullRom, select, interpolateRainbow} from 'd3'
import {real} from './data'
import {layerings, decrossings, coords, sources, layouts} from './d3dag'
function isCmd(d) {
  return d.data.group === 'cmd'
}
export default ({
  width,
  height,
  rmax,
  config,
}: {
  width: number,
  height: number,
  rmax: number,
  config: *,
}) =>
  initFull(real, {
    layout: {
      layout: layouts[config.layout],
      size: {width: width - rmax, height: height - rmax},
      layering: layerings[config.layering],
      decross: decrossings[config.decrossing],
      coord: coords[config.coords],
    },
    sources: sources.grafo,
    interpolate: interpolateRainbow,
    line: {
      curve: curveCatmullRom,
      x: d => d.x,
      y: d => d.y,
    },
    nodeBackground: colorMap => ({
      height: d => (isCmd(d) ? 8 : 2) * 2,
      width(d) {
        const text = nodeTextSwitch(d.data.type)
        const symbols = text.length
        return symbols * 4 + 8
      },
      fill(d) {
        return colorMap[d.id]
      },
      x(d) {
        const width = (() => {
          const text = nodeTextSwitch(d.data.type)
          const symbols = text.length
          return symbols * 4 + 8
        })()
        return -width / 2
      },
      y: -8,
    }),
    text: {
      text(d) {
        if (isCmd(d)) return nodeTextSwitch(d.data.type)
        return ''
      },
      fontSize: '10px',
      fontWeight: '400',
      fontFamily: 'Helvetica',
      textAnchor: 'middle',
      alignment: 'middle',
      fill: 'black',
    },
    stroke: {
      fill: 'none',
      width: 3,
    },
  })
function nodeTextSwitch(type) {
  switch (type) {
    case 'update':
      return 'upd'
    case 'compute':
      return 'f(x)'
    default:
      return type
  }
}
function initLayout({layout, size, layering, decross, coord}) {
  return layout()
    .size([size.width, size.height])
    .layering(layering)
    .decross(decross)
    .coord(coord)
}
function initEdge({curve, x, y}) {
  return lined3()
    .curve(curve)
    .x(x)
    .y(y)
}
function initColorMap({dag, interpolate}) {
  const steps = dag.size()
  const colorMap = {}
  dag.each((node, i) => {
    colorMap[node.id] = interpolate(i / steps)
  })
  return colorMap
}
function initNodes({dag, line, colorMap, stroke}) {
  const svgSelection = select('svg')
  const defs = svgSelection.append('defs') // For gradients

  svgSelection
    .append('g')
    .selectAll('path')
    .data(dag.links())
    .enter()
    .append('path')
    .attr('d', ({data}) => line(data.points))
    .attr('fill', stroke.fill)
    .attr('stroke-width', stroke.width)
    .attr('stroke', ({source, target}) => {
      const gradId = `${source.id}-${target.id}`
      const grad = defs
        .append('linearGradient')
        .attr('id', gradId)
        .attr('gradientUnits', 'userSpaceOnUse')
        .attr('x1', source.x)
        .attr('x2', target.x)
        .attr('y1', source.y)
        .attr('y2', target.y)
      grad
        .append('stop')
        .attr('offset', '0%')
        .attr('stop-color', colorMap[source.id])
      grad
        .append('stop')
        .attr('offset', '100%')
        .attr('stop-color', colorMap[target.id])
      return `url(#${gradId})`
    })

  return svgSelection
    .append('g')
    .selectAll('g')
    .data(dag.descendants())
    .enter()
    .append('g')
    .attr('transform', ({x, y}) => `translate(${x}, ${y})`)
}
function initFull(data, cfg) {
  const layout = initLayout(cfg.layout)
  const dag = cfg.sources(data)
  layout(dag)
  const colorMap = initColorMap({dag, interpolate: cfg.interpolate})
  const line = initEdge(cfg.line)
  const nodes = initNodes({dag, line, colorMap, stroke: cfg.stroke})
  addNodeBackground(nodes, cfg.nodeBackground(colorMap))
  addTextNode(nodes, cfg.text)
}

function addNodeBackground(nodes, {width, height, fill, x, y}) {
  nodes
    .append('rect')
    .attr('width', width)
    .attr('height', height)
    .attr('fill', fill)
    .attr('x', x)
    .attr('y', y)
}

function addTextNode(
  nodes,
  {text, fontSize, fontWeight, fontFamily, textAnchor, alignment, fill},
) {
  nodes
    .append('text')
    .text(text)
    .attr('font-weight', fontWeight)
    .attr('font-size', fontSize)
    .attr('font-family', fontFamily)
    .attr('text-anchor', textAnchor)
    .attr('alignment-baseline', alignment)
    .attr('fill', fill)
}
