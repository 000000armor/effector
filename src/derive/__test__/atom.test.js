//@flow

import {atom, withEquality} from '..'

import dag from '../../graph'

class RW<A> {
 value: A
 refs: Array<(A) => void> = []
 constructor(value: A) {
  this.value = value
 }
 get(): A {
  return this.value
 }
 set(value: A) {
  this.value = value
 }
 watch(fn: (_: A) => void) {
  this.refs.push(fn)
 }
}

type SetTask<A> = {
 type: 'set',
 value: A,
 target: RW<A>,
}

type ComputeTask<A> = {
 type: 'compute',
 value: A => A,
 target: RW<A>,
 parent: RW<A>,
}

test('with graph', async() => {
 const val1 = atom(0)
 const val2 = val1.map(n => {
  const nn = n + 1
  return nn
 })
 const v3 = new RW(0)
 const v4 = new RW(-1)
 const setV3: SetTask<number> = {
  type: 'set',
  value: 3,
  target: v3,
 }
 const updV4: ComputeTask<number> = {
  type: 'compute',
  value(n) {
   const nn = n + 1
   return nn
  },
  target: v4,
  parent: v3,
 }
 const upd = await dag({
  graph: new Map([[setV3, []], [updV4, [setV3]]]),
  async task<A>(task: SetTask<A> | ComputeTask<A>) {
   switch (task.type) {
    case 'set': {
     task.target.set(task.value)
     return task.value
    }
    case 'compute': {
     const n = task.parent.get()
     const result = task.value(n)
     task.target.set(result)
     return result
    }
   }
  },
 })
 expect(upd).toMatchSnapshot()
 expect(v4.get()).toBe(4)
})

test('can be dereferenced via .get to obtain its current state', () => {
 expect(atom(0).get()).toBe(0)
})

test('can be .set to change its current state', () => {
 const n = atom(0)
 n.set(1)
 expect(n.get()).toBe(1)
})

test('can .update (a la swap in clojure)', () => {
 const n = atom(0)
 n.set(1)
 const double = x => x * 2
 n.update(double)
 expect(n.get()).toBe(2)
 n.update(double)
 expect(n.get()).toBe(4)

 expect(() => {
  //$off expect error
  n.update()
 }).toThrow()
})

test('update with arguments', () => {
 const n = atom(0)
 const sum = (value, ...args) => args.reduce((acc, d) => acc + d, value)
 n.update(sum, 1)
 expect(n.get()).toBe(1)
 n.update(sum, 1, 2)
 expect(n.get()).toBe(4)
 n.update(sum, 1, 2, 3)
 expect(n.get()).toBe(10)
 n.update(sum, 1, 2, 3, 4)
 expect(n.get()).toBe(20)
 expect(() => {
  //$off expect error
  n.update(sum, 1, 2, 3, 4, 5)
  expect(n.get()).toBe(35)
 }).not.toThrow()
})

test('can include an equality-checking function', () => {
 const a = atom(0)
 const b = withEquality(a, () => false)
 // creates a brand new atom
 expect(a).not.toBe(b)

 let numReactions = 0
 a.react(
  () => {
   numReactions++
  },
  {skipFirst: true},
 )
 b.react(
  () => {
   numReactions++
  },
  {skipFirst: true},
 )

 expect(numReactions).toBe(0)
 a.set(0)
 expect(numReactions).toBe(0)
 a.set(0)
 expect(numReactions).toBe(0)

 b.set(0)
 expect(numReactions).toBe(1)
 b.set(0)
 expect(numReactions).toBe(2)
})

test('only likes functions for equality functions', () => {
 expect(() => {
  //$off
  withEquality(atom(4), 'yo')
 }).toThrow()
 expect(() => {
  //$off
  withEquality(atom(4), 0)
 }).toThrow()
})

test('the concurrent modification of _reactors bug doesnt happen any more', () => {
 const $A = atom(false)
 const $B = atom(false)

 let A_success = false
 let C_success = false

 $A.react(
  () => {
   A_success = true
  },
  {from: $A},
 )

 const $C = $A.map(a => a && $B.get())

 $C.react(
  () => {
   C_success = true
  },
  {from: $C},
 )

 expect(A_success).toBe(false)
 expect(C_success).toBe(false)
 // used to be that this would cause the 'from' controller on C to be ignored
 // during the runReactor iteration in .set
 $A.set(true)
 expect(A_success).toBe(true)
 expect(C_success).toBe(false)
 $B.set(true)

 expect(C_success).toBe(true)
})
