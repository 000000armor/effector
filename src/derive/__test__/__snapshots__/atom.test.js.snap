// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with graph 1`] = `
Object {
  "safe": true,
  "values": Map {
    Object {
      "target": RW {
        "refs": Array [],
        "value": 3,
      },
      "type": "set",
      "value": 3,
    } => 3,
    Object {
      "parent": RW {
        "refs": Array [],
        "value": 3,
      },
      "target": RW {
        "refs": Array [],
        "value": 4,
      },
      "type": "compute",
      "value": [Function],
    } => 4,
  },
}
`;
